File: Search_in_Rotated_Sorted_Array.java
public class Solution {
    public int search(int[] nums, int target) {
        if(nums.length == 1) return nums[0] == target ? 0 : -1;
        int left = -1;
        int right = -1;
        for(int i=0; i+1<nums.length; i++) {
            if(nums[i] > nums[i+1]) {
                left = bsearch(nums, target, 0, i);
                right = bsearch(nums, target, i+1, nums.length-1);
                if(left == -1 && right == -1) return -1;
                else if(left == -1) return right;
                else if(right == -1) return left;
                return -1;
            }
        }
        return bsearch(nums, target, 0, nums.length-1);
    }
    
    private int bsearch(int[] n, int target, int left, int right) {
        while(left <= right) {
            int mid = (left + right) / 2;
            if(n[mid] == target) return mid;
            if(n[mid] < target) left = mid + 1;
            else right = mid - 1;
        }
        return -1;
    }
}

File: Swift/search-in-rotated-sorted-array.swift
/**
 * https://leetcode.com/problems/search-in-rotated-sorted-array/
 * 
 * 
 */ 

class Solution {
    ///
    /// - Complexity: O(logn)
    ///
    func search(_ nums: [Int], _ target: Int) -> Int {
        /// 1. find the 0th index of number in nums.
        /// Using binary search to find the index of very first element, which should be the smallest element.
        var left = 0
        var right = nums.count - 1
        while left < right {
            let mid = left + (right - left) / 2
            if nums[mid] > nums[right] {
                left = mid + 1
            } else {
                right = mid
            }
        }
        
        let offset = left
        print("Offset: \(offset)")
        
        /// 2. find the index of target.
        /// Using binary search to find target in sorted array.
        /// Trick here is to find the index with offset.
        left = 0
        right = nums.count - 1
        while left <= right {
            let mid = left + (right - left) / 2
            let midWithOffset = (mid + offset) % nums.count
            if nums[midWithOffset] == target {
                return midWithOffset
            } else if nums[midWithOffset] > target {
                right = mid - 1
            } else {
                left = mid + 1
            }
        }
        return -1
    }
}
